### PR TITLE
fabric-loot-api-v3

### PR DESCRIPTION
**Pull request by modmuss50**
Superseeds https://github.com/FabricMC/fabric/pull/3847 by creating a V3 module passing the registry context to the events.

The V2 API has been implemented ontop of the V3 API.

Before merging we need to confirm that this is backwards compatible with the interface injection.

### CODE REVIEW

File: fabric-loot-api-v3/src/main/java/net/fabricmc/fabric/api/loot/v3/LootTableEvents.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	 * We'll add a new diamond {@linkplain net.minecraft.loot.LootPool loot pool} to the cobblestone loot table
+	 * that will be dropped alongside the original cobblestone loot pool.
+	 *
+	 * <p>If you want only one of the items to drop, you can use
+	 * {@link FabricLootTableBuilder#modifyPools(java.util.function.Consumer)} to add the new item to
+	 * the original loot pool instead.
+	 *
+	 * <pre>
+	 * {@code
+	 * LootTableEvents.MODIFY.register((key, tableBuilder, source) -> {
```
Comment (Juuxel) [inline fabric-loot-api-v3/src/main/java/net/fabricmc/fabric/api/loot/v3/LootTableEvents.java]:
```suggestion
	 * LootTableEvents.MODIFY.register((key, tableBuilder, source, registries) -> {
```

File: fabric-loot-api-v3/src/testmod/java/net/fabricmc/fabric/test/loot/LootTest.java
Comment (Juuxel) [inline fabric-loot-api-v3/src/testmod/java/net/fabricmc/fabric/test/loot/LootTest.java:1]: The modifications here and in the v2 test seem to be the same so the automated tests only make sure that one of them works. I guess that's fine, considering the v2 impl is now a direct and very basic bridge to v3.

File: fabric-loot-api-v3/src/main/java/net/fabricmc/fabric/api/loot/v3/LootTableEvents.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	 * <h2>Example: adding diamonds to the cobblestone loot table</h2>
+	 * We'll add a new diamond {@linkplain net.minecraft.loot.LootPool loot pool} to the cobblestone loot table
+	 * that will be dropped alongside the original cobblestone loot pool.
+	 *
+	 * <p>If you want only one of the items to drop, you can use
+	 * {@link FabricLootTableBuilder#modifyPools(java.util.function.Consumer)} to add the new item to
+	 * the original loot pool instead.
+	 *
+	 * <pre>
+	 * {@code
```
Comment (Juuxel) [inline fabric-loot-api-v3/src/main/java/net/fabricmc/fabric/api/loot/v3/LootTableEvents.java]: We could use `{@snippet}` here now that it's available for slightly cleaner code (and IDEA automatically links some of the APIs used when you look at the javadoc there)

File: fabric-loot-api-v3/src/main/java/net/fabricmc/fabric/api/loot/v3/LootTableEvents.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	/**
+	 * This event can be used for post-processing after all loot tables have been loaded and modified by Fabric.
+	 */
+	public static final Event<Loaded> ALL_LOADED = EventFactory.createArrayBacked(Loaded.class, listeners -> (resourceManager, lootManager) -> {
+		for (Loaded listener : listeners) {
+			listener.onLootTablesLoaded(resourceManager, lootManager);
+		}
+	});
+
+	public interface Replace {
```
Comment (maityyy) [inline fabric-loot-api-v3/src/main/java/net/fabricmc/fabric/api/loot/v3/LootTableEvents.java:100]: Missed `@FunctionalInterface`

File: fabric-loot-api-v3/src/testmod/java/net/fabricmc/fabric/test/loot/LootTest.java
Comment (modmuss50) [reply fabric-loot-api-v3/src/testmod/java/net/fabricmc/fabric/test/loot/LootTest.java:1]: As the bridge is so basic I think ill just change the V2 tests to make sure they are invoked 👍

Comment (modmuss50) [reply fabric-loot-api-v3/src/testmod/java/net/fabricmc/fabric/test/loot/LootTest.java:1]: On 2nd thoughts its easier to just change the items/blocks the test uses.

**Pull request merged**